libraries/chain/controller.cpp

```diff
@@ -3231,23 +3292,70 @@ const flat_set<account_name> &controller::get_resource_greylist() const {
 }
 
 
-void controller::add_to_ram_correction( account_name account, uint64_t ram_bytes ) {
+void controller::add_to_ram_correction( account_name account, uint64_t ram_bytes, uint32_t action_id, const char* event_id ) {
+   int64_t correction_object_id = 0;
+
    if( auto ptr = my->db.find<account_ram_correction_object, by_name>( account ) ) {
       my->db.modify<account_ram_correction_object>( *ptr, [&]( auto& rco ) {
+         correction_object_id = rco.id._id;
          rco.ram_correction += ram_bytes;
       } );
    } else {
       my->db.create<account_ram_correction_object>( [&]( auto& rco ) {
+         correction_object_id = rco.id._id;
          rco.name = account;
          rco.ram_correction = ram_bytes;
       } );
    }
+
+   if (auto dm_logger = get_deep_mind_logger()) {
+      fc_dlog(*dm_logger, "RAM_CORRECTION_OP ${action_id} ${correction_id} ${event_id} ${payer} ${delta}",
+         ("action_id", action_id)
+         ("correction_id", correction_object_id)
+         ("event_id", event_id)
+         ("payer", account)
+         ("delta", ram_bytes)
+      );
+   }
 }
 
 bool controller::all_subjective_mitigations_disabled()const {
    return my->conf.disable_all_subjective_mitigations;
 }
 
+fc::logger* controller::get_deep_mind_logger()const {
+   return my->get_deep_mind_logger();
+}
+
+void controller::enable_deep_mind(fc::logger* logger) {
+   EOS_ASSERT( logger != nullptr, misc_exception, "Invalid logger passed into enable_deep_mind, must be set" );
+   my->deep_mind_logger = logger;
+
+   // The actual `fc::dmlog_appender` implementation that is currently used by deep mind
+   // logger is using `stdout` to prints it's log line out. Deep mind logging outputs
+   // massive amount of data out of the process, which can lead under pressure to some
+   // of the system calls (i.e. `fwrite`) to fail abruptly without fully writing the
+   // entire line.
+   //
+   // Recovering from errors on a buffered (line or full) and continuing retrying write
+   // is merely impossible to do right, because the buffer is actually held by the
+   // underlying `libc` implementation nor the operation system.
+   //
+   // To ensure good functionalities of deep mind tracer, the `stdout` is made unbuffered
+   // and the actual `fc::dmlog_appender` deals with retry when facing error, enabling a much
+   // more robust deep mind output.
+   //
+   // Changing the standard `stdout` behavior from buffered to unbuffered can is disruptive
+   // and can lead to weird scenarios in the logging process if `stdout` is used there too.
+   //
+   // In a future version, the `fc::dmlog_appender` implementation will switch to a `FIFO` file
+   // approach, which will remove the dependency on `stdout` and hence this call.
+   //
+   // For the time being, when `deep-mind = true` is activated, we set `stdout` here to
+   // be an unbuffered I/O stream.
+   setbuf(stdout, NULL);
+}
```

Comment:
Move this to `dmlog.hpp` as a standalone function.

Reply:
The whole `enable_...` function or just setting `stdout` unbuffered?

Reply:
Well, now that controller still has a logger pointer, just the large comment and `setbuf`

Reply:
Added `set_dmlog_appender_stdout_unbuffered` in a `deep_mind.hpp` file.